main/src/ca/uwaterloo/flix/tools/BenchmarkCompiler.scala

```diff
@@ -1,226 +1,176 @@
+import ca.uwaterloo.flix.runtime.CompilationResult
 import ca.uwaterloo.flix.util.{LocalResource, Options, StatUtils, Validation}
 import ca.uwaterloo.flix.util.vt.TerminalContext
+import org.json4s.JsonDSL._
+import org.json4s.native.JsonMethods
 
 /**
   * A collection of internal utilities to measure the performance of the Flix compiler itself.
   */
 object BenchmarkCompiler {
 
-  /**
-    * The number of compilations to perform before the statistics are collected.
-    */
-  val WarmupIterations = 25
-
   /**
     * The number of compilations to perform when collecting statistics.
     */
-  val BenchmarkIterations = 10
+  val N = 15
 
   /**
     * Outputs statistics about time spent in each compiler phase.
     */
   def benchmarkPhases(o: Options): Unit = {
-    val flix = newFlix(o)
-
-    warmup(flix)
-
-    val results = List.fill(BenchmarkIterations) {
-      flix.compile() match {
-        case Validation.Success(_) =>
-          flix.phaseTimers.toList
-        case Validation.Failure(errors) =>
-          errors.sortBy(_.source.name).foreach(e => println(e.message.fmt(TerminalContext.AnsiTerminal)))
-          System.exit(1)
-          throw new AssertionError("impossible") // inaccessible
-      }
+    //
+    // Collect data from N iterations.
+    //
+    val r = (0 until N).map {
+      case _ =>
+        val flix = newFlix(o)
+        val compilationResult = flix.compile().get
+        (compilationResult, flix.phaseTimers.toList)
     }
 
+    //
+    // Split into compilation results and phase results.
+    //
+    val results = r.map(_._1).toList
+    val phases = r.map(_._2)
+
     // phase -> list of times
-    val phaseMap = results.flatten.groupMap(_.phase)(_.time)
+    val phaseMap = phases.flatten.groupMap(_.phase)(_.time)
 
     // phase -> median time
     val phaseMedians = phaseMap.map {
-      case (phase, times) => (phase, StatUtils.median(times))
-    }
-
-    for ((phase, time) <- phaseMedians) {
-      val currentTime = System.currentTimeMillis() / 1000
-      println(s"$phase, $currentTime, $time")
+      case (phase, times) => (phase, StatUtils.median(times.toList))
+    }.toList
+
+    // The number of threads used.
+    val threads = o.threads
+
+    // Find the number of lines of source code.
+    val lines = results.head.getTotalLines().toLong
+
+    // Find the timings of each run.
+    val timings = results.map(_.getTotalTime())
+
+    // Compute the total time in seconds.
+    val totalTime = (timings.sum / 1_000_000_000L).toInt
+
+    // Print JSON or plain text?
+    if (o.json) {
+      val json =
+        ("threads" -> threads) ~
+          ("lines" -> lines) ~
+          ("iterations" -> N) ~
+          ("phases" -> phaseMedians.map {
+            case (phase, time) => ("phase" -> phase) ~ ("time" -> time)
+          })
+      val s = JsonMethods.pretty(JsonMethods.render(json))
+      println(s)
+    } else {
+      println("====================== Flix Compiler Phases ======================")
```

Comment:
Note that this will break the Github action benchmark runner.

Reply:
Yes, I know :(

Reply:
But having JSON output will be much easier to parse in the future.

And I am trying to hook it up to the Rasperry PI.